staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper_test.go

```diff
@@ -0,0 +1,366 @@
+	var flushInvoked, closeNotifyInvoked int
+	wrap := func(inner http.ResponseWriter) http.ResponseWriter {
+		decorator := &fakeResponseWriterDecorator{ResponseWriter: inner}
+		return Wrap(decorator, Callback{
+			Flush: func() {
+				flushInvoked++
+				decorator.ResponseWriter.(http.Flusher).Flush()
+			},
+			CloseNotify: func() <-chan bool {
+				closeNotifyInvoked++
+				//lint:ignore SA1019 backward compatibility
+				return decorator.ResponseWriter.(http.CloseNotifier).CloseNotify()
+			},
+		})
+	}
+
+	inner := &FakeResponseWriterFlusherCloseNotifier{}
+	wrapped := wrap(inner)
+	wrapped = wrap(wrapped)
+	wrapped = wrap(wrapped)
+
+	if _, ok := wrapped.(*withCloseNotifyAndFlush); !ok {
+		t.Errorf("Expected the wrapped http.ResponseWriter object to be of *withCloseNotifyAndFlush type, but got: %T", wrapped)
+	}
+
+	// we wrap the inner http.ResponseWriter object three times.
+	var flushExpected, closeNotifyExpected = 3, 3
+
+	//lint:ignore SA1019 backward compatibility
+	wrapped.(http.CloseNotifier).CloseNotify()
+	wrapped.(http.Flusher).Flush()
+
+	if flushExpected != flushInvoked {
+		t.Errorf("Expected Flush to be invoked %d times, but got: %d", flushExpected, flushInvoked)
+	}
+	if closeNotifyExpected != closeNotifyInvoked {
+		t.Errorf("Expected CloseNotify to be invoked %d times, but got: %d", closeNotifyExpected, closeNotifyInvoked)
+	}
+
+	unwrapped := Unwrap(wrapped)
+	if inner != unwrapped {
+		t.Errorf("Expected the unwrapped to match, want: %#v, got: %#v", inner, unwrapped)
+	}
+}
+
+func TestResponseWriterWrapWithHTTP1(t *testing.T) {
+	var originalWant http.ResponseWriter
+	chain := func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if originalWant == nil {
+				originalWant = w
+			}
+
+			// for http/1.x request we expect the http.ResponseWriter object to
+			// implement http.Flusher, http.CloseNotifier and http.Hijacker
+			if _, ok := w.(http.Flusher); !ok {
+				t.Errorf("Expected the http.ResponseWriter object to implement http.Flusher for HTTP/1.x")
+			}
+
+			//lint:ignore SA1019 backward compatibility
+			if _, ok := w.(http.CloseNotifier); !ok {
+				t.Errorf("Expected the http.ResponseWriter object to implement http.CloseNotifier for HTTP/1.x")
+			}
+			if _, ok := w.(http.Hijacker); !ok {
+				t.Errorf("Expected the http.ResponseWriter object to implement http.Hijacker for HTTP/1.x")
+			}
+
+			decorator := &fakeResponseWriterDecorator{ResponseWriter: w}
+			wrapped := Wrap(decorator, Callback{
+				Flush: func() {
+					decorator.flush()
+				},
+				CloseNotify: func() <-chan bool {
+					return decorator.closeNotify()
+				},
+				Hijack: func() (net.Conn, *bufio.ReadWriter, error) {
+					return decorator.hijack()
+				},
+			})
+			if _, ok := wrapped.(*withCloseNotifyFlushAndHijack); !ok {
+				t.Errorf("Expected the wrapped ResponseWriter object to be of withCloseNotifyFlushAndHijack type")
+				return
+			}
```

Comment:
I prefer behavioral tests over ones that simply reiterate the implementation.  We could make this a behavioral test by counting the number of method invocations, like in the first two tests.  In fact, if we count the method invocations in these tests then we do not really need to also have the first two tests.

Reply:
there were two tests now - `TestWithHTTP1` and `TestWithHTTP2`.